src/net/ossl.cc

```diff
@@ -453,5 +462,592 @@ void tls::server_credentials::set_client_auth(client_auth ca) {
+                    return make_ready_future<>();
+                });
+            }).then([]{
+                netlogger.info("SSL handshake complete");
+            });
+    }
+
+    future<> wait_for_input() {
+        if (eof()) {
+            return make_ready_future<>();
+        }
+        netlogger.info("Inside session::wait_for_input");
+        return _in.get().then([this](buf_type data) {
+            if (data.empty()) {
+                _eof = true;
+                return make_ready_future<>();
+            }
+            // Write the received data to the "read bio".  This bio is consumed
+            // by the SSL struct.  Think of this of writing encrypted data into
+            // the SSL session
+            netlogger.info("Read {} bytes over the wire", data.size());
+            const auto input_len = data.size();
+            auto buf = make_lw_shared<buf_type>(std::move(data));
+            return do_until(
+              [buf]{ return buf->empty(); },
+              [this, buf]{
+                  const auto n = BIO_write(_in_bio, buf->get(), buf->size());
+                  netlogger.info("Wrote {} bytes over to SSL in bio", n);
+                  if (n <= 0) {
+                      return make_exception_future<>(ossl_error(n,  "Error while waiting for input"));
+                  }
+                  buf->trim_front(n);
+                  return make_ready_future();
+              }).finally([buf, input_len]{
+                  netlogger.info("Wrote {} bytes into internal SSL buffers", input_len);
+              });
+        }).handle_exception([](auto ep){
+            return make_exception_future(ep);
+        });
+    }
+
+    future<buf_type> do_get() {
+        netlogger.info("Inside of session::do_get()");
+        // Check if there is encrypted data sitting in ssls internal buffers, otherwise wait
+        // for data and use a
+        auto f = make_ready_future<>();
+        auto avail = BIO_ctrl_pending(_in_bio);
+        if (avail == 0) {
+            if (eof()) {
+                return make_ready_future<buf_type>(buf_type());
+            }
+            f = wait_for_input();
+        }
+        return f.then([this]() {
+            const auto buf_size = 4096;
+            buf_type buf(buf_size);
+            // Read decrypted data from ssls internal buffers
+            auto bytes_read = SSL_read(_ssl.get(), buf.get_write(), buf_size);
+            if (bytes_read <= 0) {
+                const auto ec = SSL_get_error(_ssl.get(), bytes_read);
+                if (ec == SSL_ERROR_ZERO_RETURN && connected()) {
+                    // Client has initiated shutdown by sending EOF
+                    netlogger.info("Inside detection os SSL_ERROR_ZERO_RETURN");
+                    _eof = true;
+                    close();
+                    return make_ready_future<buf_type>(buf_type());
+                }
+                return make_exception_future<buf_type>(ossl_error(ec, "error via SSL_read"));
```

Comment:
This could also be a renegotation event here, where `SSL_get_error` could return WANT_READ or WANT_WRITE

Reply:
similar comment to above, do we want to file another ticket to handle a re-negotiate situation or handle it now?

Reply:
ticket filed, resolving